gamestonk_terminal/portfolio_optimization/port_opt_api.py

```diff
@@ -0,0 +1,360 @@
+        rets = w.dot(mu)
+        stds = np.sqrt(np.diag(w @ S @ w.T))
+        sharpes = rets / stds
+        ax.scatter(stds, rets, marker=".", c=sharpes, cmap="viridis_r")
+
+        plotting.plot_efficient_frontier(ef, ax=ax, show_assets=True)
+        # Find the tangency portfolio
+        ef.max_sharpe()
+        ret_sharpe, std_sharpe, _ = ef.portfolio_performance()
+        ax.scatter(std_sharpe, ret_sharpe, marker="*", s=100, c="r", label="Max Sharpe")
+
+        ax.set_title("Efficient Frontier")
+        ax.legend()
+        plt.tight_layout()
+        plt.grid(b=True, which="major", color="#666666", linestyle="-")
+        plt.minorticks_on()
+        plt.grid(b=True, which="minor", color="#999999", linestyle="-", alpha=0.2)
+
+        if gtff.USE_ION:
+            plt.ion()
+
+        plt.show()
+        print("")
+
+    except Exception as e:
+        print(e)
+        print("")
+
+
+def ef_portfolio(list_of_stocks: List[str], port_type: str, other_args: List[str]):
+    """
+    Return a portfolio based on condition in port_type  Defaults to 3m of historical data
+    Parameters
+    ----------
+    list_of_stocks: List[str]
+        List of the stocks to be included in the weights
+    port_type: str
+        Method to be used on ef object (example: max_sharpe, min_volatility)
+    Returns
+    -------
+    weights: dict
+        Dictionary of weights where keys are the tickers.
+    """
+
+    parser = argparse.ArgumentParser(add_help=False, prog="ef")
+
+    parser.add_argument(
+        "-p",
+        "--period",
+        default="3mo",
+        dest="period",
+        help="period to get yfinance data from",
+        choices=period_choices,
+    )
+
+    parser.add_argument(
+        "-v",
+        "--value",
+        dest="value",
+        help="Amount to allocate to portfolio",
+        type=float,
+        default=1.0,
+    )
+    parser.add_argument(
+        "--pie",
+        action="store_true",
+        dest="pie",
+        default=False,
+        help="Display a pie chart for weights",
+    )
+    if port_type == "max_sharpe":
+        try:
+            ns_parser = parse_known_args_and_warn(parser, other_args)
+            if not ns_parser:
+                return {}
+            period = ns_parser.period
+            stock_prices = process_stocks(list_of_stocks, period)
+            ef = prepare_efficient_frontier(stock_prices)
+            ef_sharpe = dict(ef.max_sharpe())
+            weights = {
+                key: ns_parser.value * round(value, 5)
+                for key, value in ef_sharpe.items()
+            }
+            if ns_parser.pie:
+                pie_chart_weights(weights)
+            print("")
+            ef.portfolio_performance(verbose=True)
+            print("")
+            return weights
+        except Exception as e:
+            print(e)
+            print("")
+            return {}
+
+    elif port_type == "min_volatility":
```

Comment:
It seems that you can optimize the code a bit further. The initial parts are the same for all, the only difference is when you get ef_min_vol, or ef_sharpe, ...  So you can just have an if there.

Reply:
You can even have a function that get this data for you based on the port_type. This would be the preferred approach